snap/info_snap_yaml_test.go

```diff
@@ -1608,6 +1608,8 @@ layout:
     bind: $SNAP/usr/share/foo
   /usr/share/bar:
     symlink: $SNAP/usr/share/bar
+  /etc/froz:
+    bind-file: $SNAP/etc/froz
```

Comment:
What happens if one does:
```
/etc/norf:
  bind-file: $SNAP/etc/norf
  bind: $SNAP/etc/norf
```
?

What happens if you have:
```
/etc/norf:
  bind: /etc/norf
/etc/norf:
  bind-file:/etc/norf
```
?

What happens if you have:
```
/etc/norf:
  bind: $SNAP/etc/norf
/etc/corge:
  bind-file: $SNAP/etc/corge
```
?

In other words, what prevents the various permutations of bind, bind-file and symlink rules for the source and/or dest from colliding? Can you add tests for these (my list wasn't exhaustive)? The info_snap_yaml_test.go only has the one positive test; it would be good to have negative tests too.

Reply:
All of those are caught by the validator. You can use at most one of the "kind" of layout item and there's extensive checks that you cannot sneak one under another.

Reply:
"This one is caught by the validator"

Sure, but where is the test? :)


"This one is more subtle:"

You can certainly detect if you've specified the same dest multiple times.... When you do, can you add a test? :)


Meh, I wrote the third example wrong. It was meant to be:
```
/etc/norf:
  bind: $SNAP/etc/norf
/etc/corge:
  bind-file: $SNAP/etc/norf
```

In other words, the source is the same but one rule says it is a Bind mount and the other a BindFile mount.

My point wasn't necessarily that the code is wrong (though based on your answer to the second question, it might be), but that there are no negative tests AFAICS. Having a symlink and a Bind with different sources pointing to the same dest is, for example, fine, but having two identical sources with one using a Bind and the other a BindFile are clearly not-- the source can only be a file or a dir, not both. We need to enumerate everything that is ok and everything that isn't, and then have positive and negative tests for all.

Reply:
>> "This one is more subtle:"
> You can certainly detect if you've specified the same dest multiple times.... When you do, can you add a test? :)

On IRC I was reminded that these are keys to a dict so keys with different values overwrite the value. Durr...